src/module/types.rs

```diff
@@ -6,22 +6,66 @@ use crate::error::Result;
 use crate::module::Module;
 use crate::parse::IndicesToIds;
 use crate::ty::{Type, TypeId, ValType};
+use std::sync::{Arc, Mutex};
 
 /// The set of de-duplicated types within a module.
+///
 #[derive(Debug, Default)]
 pub struct ModuleTypes {
-    arena: ArenaSet<Type>,
+    arena: Mutex<ArenaSet<Type>>,
```

Comment:
Did this basically just get way too unwieldy unless we had interior mutability?

Reply:
It became unweildy when blocks were storing their param and result types directly, and then doing a get-or-create for those param/result types during emit time. But I evolved the design since then, and `InstrSeq`s hold a `TypeId` directly now. They still do a get-or-create during parsing, which happens (can happen) in parallel for each function, so the synchronization is still required. However, it would be easy to switch that into a fallible `try-and-get-the-existing-type-with-these-params-and-results` method that returns an error instead of inserting when it doesn't already exist. This should be fine since the blocks reference an existing type in the binary format.

I'll try it out!

Reply:
Ah ok, seems fine to me either way. It's sort of a bummer to require synchronization now but it's not really the end of the world.

Reply:
Removed the serialization at the cost of complicating `Type` with a boolean for whether or not it is for function entry blocks (which are implicit and therefore don't actually hold onto a `Type` in the serialization). I think this is the better choice though, and a bunch of stuff got simplified.